### PR TITLE
Fixed PR #523, prevent unnecessary saving of status history when orde…

### PR DESCRIPTION
**Pull request by kiatng**
…r is saved.

### Description (*)
Order is saved when comments are added in the backend. Before this PR, this will save the new comment (status history object) along with all the old status histories which are unnecessary. After this PR, only the new comment is saved.

### Related Pull Requests
PR #523

### Manual testing scenarios (*)

To test this, add test code in `app\code\core\Mage\Sales\Model\Order\Status\History.php`

```php
    public function save()
    {
        Mage::helper('clog')->log("sh#{$this->getId()}"); // use your own log
        return parent::save();
    }
   //...
    protected function _beforeSave()
    {
        Mage::helper('clog')->log("bsh#{$this->getId()}"); // use your own log
       //...
```
 Test result before PR:
```
[25-03-2022 15:59:52][kiat Administrators] sh#5774133
[25-03-2022 15:59:52][kiat Administrators] bsh#5774133
[25-03-2022 15:59:52][kiat Administrators] sh#5774132
[25-03-2022 15:59:52][kiat Administrators] bsh#5774132
[25-03-2022 15:59:52][kiat Administrators] sh#5774131
[25-03-2022 15:59:52][kiat Administrators] bsh#5774131
[25-03-2022 15:59:52][kiat Administrators] sh#5773982
[25-03-2022 15:59:52][kiat Administrators] bsh#5773982
[25-03-2022 15:59:52][kiat Administrators] sh#
[25-03-2022 15:59:52][kiat Administrators] bsh#
```
Test result after PR:
```
[25-03-2022 16:00:49][kiat Administrators] sh#5774134
[25-03-2022 16:00:49][kiat Administrators] sh#5774133
[25-03-2022 16:00:49][kiat Administrators] sh#5774132
[25-03-2022 16:00:49][kiat Administrators] sh#5774131
[25-03-2022 16:00:49][kiat Administrators] sh#5773982
[25-03-2022 16:00:49][kiat Administrators] sh#
[25-03-2022 16:00:49][kiat Administrators] bsh#
```

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [x] Add yourself to contributors list

### CODE REVIEW

File: app/code/core/Mage/Sales/Model/Order.php [hunk 1]
```diff
@@ -1895,8 +1895,10 @@ public function getStatusHistoryCollection($reload = false)
                 ->setOrder('entity_id', 'desc');
 
             if ($this->getId()) {
+                /** @var Mage_Sales_Model_Order_Status_History $status */
                 foreach ($this->_statusHistory as $status) {
                     $status->setOrder($this);
+                    $status->setDataChanges(false);
```
Comment (fballiano) [inline app/code/core/Mage/Sales/Model/Order.php:1901]: it seems we need to add this setDataChanges calls in a lot of places, but doesn't it mean we should modify some Abstract or something like that to solve this issue in a more general way?

**Review by tobihille — APPROVED** [review]

**Comment by kiatng** [comment]
@fballiano is right, there should be a general way to solve this, but I have not found the way to do it.

File: app/code/core/Mage/Sales/Model/Order.php [hunk 2]
```diff
@@ -1895,8 +1895,10 @@ public function getStatusHistoryCollection($reload = false)
                 ->setOrder('entity_id', 'desc');
 
             if ($this->getId()) {
+                /** @var Mage_Sales_Model_Order_Status_History $status */
```
Comment (sreichel) [inline app/code/core/Mage/Sales/Model/Order.php:1898]: Not required.

**Review by sreichel — APPROVED** [review]

File: app/code/core/Mage/Sales/Model/Order.php [hunk 3]
```diff
@@ -1895,8 +1895,10 @@ public function getStatusHistoryCollection($reload = false)
...
                 foreach ($this->_statusHistory as $status) {
```
Comment (sreichel) [inline app/code/core/Mage/Sales/Model/Order.php:1899]:
```suggestion
            if ($this->getId()) {
                foreach ($this->_statusHistory as $status) {
```

Comment (kiatng) [reply app/code/core/Mage/Sales/Model/Order.php:1899]: Without the docblock, my VSCode, says it's `mixed`:
image
So how?

Comment (sreichel) [reply app/code/core/Mage/Sales/Model/Order.php:1899]: Then we just keep it.

PhpStorm ...

pr-2042

**Comment by elidrissidev** [comment]
While the current change will work fine, I found that all collections suffer from this issue! This could be fixed in all places by applying the below patch to `lib/Varien/Data/Collection/Db.php`, and keeping the change from the original PR to prevent `setStoreId()` call from changing `_hasDataChanges` to `true` again (tested).

```patch
diff --git a/lib/Varien/Data/Collection/Db.php b/lib/Varien/Data/Collection/Db.php
index 147f25ca3d..ff85dcaada 100644
--- a/lib/Varien/Data/Collection/Db.php
+++ b/lib/Varien/Data/Collection/Db.php
@@ -587,6 +587,7 @@ class Varien_Data_Collection_Db extends Varien_Data_Collection
                     $item->setIdFieldName($this->getIdFieldName());
                 }
                 $item->addData($row);
+                $item->setDataChanges(false);
                 $this->addItem($item);
             }
         }
```

What do you think?

**Comment by kiatng** [comment]
@elidrissidev It's a good idea. Can you create a PR and close this one?

**Comment by elidrissidev** [comment]
@kiatng Sure. Will do it as soon as I get some time.

**Pull request closed without merging**